### PR TITLE
redis: add support for redis 6 acls via the `AUTH username password` command

### LINKED ISSUE #11296: Add support for Redis 6 ACL
*Title*: Add support for Redis 6 ACL

*Description*:
Redis 6 introduces ACLs which require a username and password.   We would like to continue to use Envoy as a proxy for these deployments.

[Redis ACL]https://redis.io/topics/acl

I think there are two ways to go on the implementation.   Adding an optional username alongside the password in the options, or adding a new ACL type of configuration which requires a username and a password.   Interested in your thoughts and happy to contribute.

### PR DESCRIPTION
**Pull request by jamesbibby**
Commit Message: Add support for the new Redis 6 ACL auth (AUTH username password) 
Additional Description: Adds an optional username to config and auth commands, and tests
Risk Level: Medium
Testing: Added a number of unit tests, am working on building a docker image locally to test in the wild
Docs Changes: Docs will need to be updated to show the new configuration and support for the AUTH username password command and the downstream_username configuration element
Release Notes:  Adds support for redis 6 acls via the `AUTH username password` command
Fixes #11296

### CODE REVIEW

**Comment by jamesbibby** [comment]
I still have not tested this past writing unit tests, working on building an image now but I wanted to get some feedback on the approach as it is my first PR to envoy.   I don't know the codebase well so please let me know if things make sense or if I can improve them.

Also I modified both the v2 and v3 apis for redis to include the downstream_auth_username.   Not sure if this was necessary or if just v3 should have been updated.

**Comment by mattklein123** [comment]
@HenryYYang PTAL

**Comment by jamesbibby** [comment]
I added ACL to the dictionary so the spelling tests would pass, also cleaned up some error messages/comments.

File: api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto [hunk 1]
```diff
@@ -24,7 +24,7 @@ option (udpa.annotations.file_status).package_version_status = FROZEN;
 // Redis Proxy :ref:`configuration overview <config_network_filters_redis_proxy>`.
 // [#extension: envoy.filters.network.redis_proxy]
 
-// [#next-free-field: 7]
+// [#next-free-field: 8]
```
Comment (htuch) [inline api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto:27]: Since the v2 API is frozen, can you make these changes only to the v3 APIs?

Comment (jamesbibby) [reply api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto:27]: Yes I can give that a go today.

Comment (jamesbibby) [reply api/envoy/config/filter/network/redis_proxy/v2/redis_proxy.proto:27]: @htuch The one question I have is around test.   The tests in these two files are based on a v2 config file format:
github.com/envoyproxy/envoy/blob/master/test/extensions/filters/network/redis_proxy/redis_proxy_integration_test.cc
github.com/envoyproxy/envoy/blob/master/test/extensions/clusters/redis/redis_cluster_integration_test.cc

Do these need new tests that are based on a v3 config format or are these tests only relevant to the v2 api?

**Comment by jamesbibby** [comment]
@mattklein123 I believe the mac failures were not related to this PR, is there a way to re-trigger the CI or should I push something trivial to kick things off again?

**Comment by mattklein123** [comment]
Let's wait on further pushes until @HenryYYang reviews. @HenryYYang PTAL.

File: source/extensions/filters/network/redis_proxy/proxy_filter.cc [hunk 2]
```diff
@@ -96,6 +99,25 @@ void ProxyFilter::onAuth(PendingRequest& request, const std::string& password) {
   request.onResponse(std::move(response));
 }
 
+void ProxyFilter::onAuth(PendingRequest& request, const std::string& username,
+                         const std::string& password) {
+  Common::Redis::RespValuePtr response{new Common::Redis::RespValue()};
+  if (config_->downstream_auth_username_.empty() && config_->downstream_auth_password_.empty()) {
```
Comment (HenryYYang) [inline source/extensions/filters/network/redis_proxy/proxy_filter.cc:105]: I think we need to handle the case where password is config but username is empty.  In this case the "AUTH default <password>" should return OK.

File: source/extensions/filters/network/redis_proxy/proxy_filter.cc [hunk 3]
```diff
@@ -96,6 +99,25 @@ void ProxyFilter::onAuth(PendingRequest& request, const std::string& password) {
...
+    response->type(Common::Redis::RespType::Error);
+    response->asString() = "ERR Client sent AUTH, but no ACL is set";
+  } else if (username == config_->downstream_auth_username_ &&
+             password == config_->downstream_auth_password_) {
+    response->type(Common::Redis::RespType::SimpleString);
+    response->asString() = "OK";
+    connection_allowed_ = true;
+  } else {
+    response->type(Common::Redis::RespType::Error);
+    response->asString() = "ERR invalid ACL auth";
```
Comment (HenryYYang) [inline source/extensions/filters/network/redis_proxy/proxy_filter.cc:115]: Let's use "WRONGPASS invalid username-password pair" like what Redis 6 returns.

**Comment by jamesbibby** [comment]
@HenryYYang I fixed some code/format issues to get the CI to pass, added documentation and addressed code review comments.  Should be ready for another look.

@htuch I updated the docs in the API, so I think you may need to give another lgtm for the API. (I updated the expected error when an ACL auth call is incorrect to match the redis 6 response)

File: api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto [hunk 4]
```diff
@@ -234,6 +234,18 @@ message RedisProxy {
   // client. If an AUTH command is received when the password is not set, then an "ERR Client sent
   // AUTH, but no password is set" error will be returned.
   config.core.v3.DataSource downstream_auth_password = 6 [(udpa.annotations.sensitive) = true];
+
+  // If a username is provided an ACL style AUTH command will be required with a username and password.
+  // Authenticate Redis client connections locally by forcing downstream clients to issue a `Redis
+  // AUTH command <https://redis.io/commands/auth>`_ with this username and the downstream_auth_password
```
Comment (htuch) [inline api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto:240]: Nit: for fields such as `downstream_auth_password`, either add a `:ref:` or write as `*downstream_auth_password*` if they are proximal (which is the case here).

Comment (jamesbibby) [reply api/envoy/extensions/filters/network/redis_proxy/v3/redis_proxy.proto]: Fixed.

**Comment by htuch** [comment]
@HenryYYang LMK when you think this is ready to merge, thanks.

**Comment by HenryYYang** [comment]
👍

**Comment by mattklein123** [comment]
Sorry needs a master merge.

/wait

**Comment by jamesbibby** [comment]
Do you want me to merge master or will you folks take care of that?

**Comment by mattklein123** [comment]
You need to merge master, thank you.

**Comment by jamesbibby** [comment]
> You need to merge master, thank you.

Done and pushed

**Comment by jamesbibby** [comment]
Fixed a merge issue (comments out of alpha order)

**Comment by jamesbibby** [comment]
Thanks for working with me on this.   My first serious c++ contribution in like 2 decades :)

**Pull request merged**